### PR TITLE
Add GitHub Actions workflow

### PR DESCRIPTION
**Pull request by tothtamas28**
~Blocked on: #7~

### CODE REVIEW

File: {{cookiecutter.project_slug}}/.github/workflows/test-pr.yml [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+          docker run                 \
+            --name ${CONTAINER}      \
+            --rm                     \
+            --interactive            \
+            --tty                    \
+            --detach                 \
+            ${IMAGE}
+
+          docker cp . ${CONTAINER}:/home/user
+          docker exec ${CONTAINER} chown -R user:user /home/user
```
Comment (ehildenb) [inline {{cookiecutter.project_slug}}/.github/workflows/test-pr.yml:53]: We should pull this out into an action, so that it's easier for users to set up docker for mulitple different stages if their project calls for that. Here is an [example of a similar action](https://github.com/runtimeverification/k/blob/master/.github/actions/with-docker/action.yml), and [how it is used here](https://github.com/runtimeverification/k/blob/master/.github/workflows/test-pr.yml#L85)

If you want to leave this for a later PR that is OK too, but in terms of starting out repos from a better starting point I think this will help.

Freeman and I are still working on getting rid of the need for `docker exec ...` prefixing each test line too.

Comment (ehildenb) [reply {{cookiecutter.project_slug}}/.github/workflows/test-pr.yml:53]: Also, @F-WRunTime look how here we are using `docker cp . ${CONTAINER}:/home/user` followed by `docker exec ${CONTAINER} chown -R user:user /home/user`. I think this could be useful other places, instead of the volume mounting technique we're doing.

Benefit is that you know your environment will be cleaner each time, and the copied in data is not persistent.

Drawback is that the copied in data is not persistent, so pushing artifacts around between builds  is probably harder.

File: {{cookiecutter.project_slug}}/.github/workflows/test-pr.yml [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+jobs:
+  code-quality-checks:
+    name: 'Code Quality Checks'
+    runs-on: ubuntu-latest
+    steps:
+      - name: 'Check out code'
+        uses: actions/checkout@v3
+      - name: 'Install dependencies'
+        run: |
```
Comment (ehildenb) [inline {{cookiecutter.project_slug}}/.github/workflows/test-pr.yml:17]: We should put a big warning here: "IF YOU ARE ON A PRIVATE REPO, SWITCH THIS TO PRIVATE RUNNER", otherwise when people copy this template they will very quickly wrack up github bills.

File: {{cookiecutter.project_slug}}/.github/workflows/test-pr.yml [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+        os: [ubuntu-latest, macos-11]
+    runs-on: {% raw %}${{ matrix.os }}{% endraw %}
+    steps:
+      - name: 'Check out code'
+        uses: actions/checkout@v3
+      - name: 'Install Nix'
+        uses: cachix/install-nix-action@v15
+        with:
+          extra_nix_config: |
+            substituters = http://cache.nixos.org https://cache.iog.io
```
Comment (ehildenb) [inline {{cookiecutter.project_slug}}/.github/workflows/test-pr.yml:75]: Similarly here, we shoudl have a warning.

Maybe even better is to have the template default to private runners only, and then put a comment saying they can switch to public if it's a public repo.

File: {{cookiecutter.project_slug}}/.github/workflows/test-pr.yml (hunk 1 above)
Comment (FWBlueWizard) [reply {{cookiecutter.project_slug}}/.github/workflows/test-pr.yml:53]: As well as publishing anything from a workflow. BUT as we've been using nix and all that publishing artifacts via workflow artifact pushes to github has become less so necessary. As well as introducing pypi. 

SO I agree I don't have a hardline at this time on mounts. As we've been moving further from need to publish artifacts the host runner & action needs read/write access to to push. 
This did  however introduce permission issues still with copying files into a container that had no knowledge of the user that copied them in. 
HOWEVER Combine the cp method with our method of building containers with the provided knowledge of the running host user UID/GID into the container, I think we hit a sweet spot of proper user UID information and internal container clean-state. 

Let me know if that was too much ramble of current progress we've made with these container builds. 

**TLDR** I agree, lets make sure we build the containers and create this basic 'user'  with the UID/GID of our non-permissive executing 'github-runner' from using `id -u` and `id -g` and we're in a good place to avoid obscure permission issues.

File: {{cookiecutter.project_slug}}/.github/workflows/test-pr.yml (hunk 2 above)
Comment (tothtamas28) [reply {{cookiecutter.project_slug}}/.github/workflows/test-pr.yml:17]: I feel adding these CI scripts to the template is not the right solution. Python and CI are two separate concerns, and we should handle them as such.

So instead, we could ether
a) Create a repo with well-documented scripts, actions, etc. that downstream projects can import.
b) Create a repo with a CI template. We can include more complex logic to control what sort of scripts are generated (volume vs cp, public vs private, Docker vs host, etc.).

Then the Python template can still be used without committing to the way we do CI in our infrastructure.

Comment (ehildenb) [reply {{cookiecutter.project_slug}}/.github/workflows/test-pr.yml:17]: Hmmmmm, you have a good point there. Maybe this belongs in devops or devops-actions repo instead. For now, leave this as a draft until I can discuss more with @F-WRunTime .
